### ISSUE TITLE
Method to Expect() error handling (recoverablity) to be triggered

### ISSUE THREAD

**Issue by ppittle** [Particular/NServiceBus.Testing#90]
`<package id="NServiceBus.Testing" version="6.0.0-beta0001" targetFramework="net452" />`

I'm writing unit tests to cover edge cases on a Handler that can partially recover from certain Exceptions, but relies on [NServiceBus Recoverability](http://docs.particular.net/nservicebus/recoverability/) for others.

```
public Task Handle(ExampleMessage message, IMessageHandlerContext context){
{
     try
     {
           DoSomething();
     }
     catch(NonTransiantException e)
    {
         context.Reply(new NonTransientExceptionMessage("No amount of retry will fix this");
    }
    catch
    {
         //use NSB recoverabiltiy to retry
        throw;
    }
}
```

It would be nice to have **Expect Methods** that could cover this:

```
Test.Handler<ExampleHandler>()
     .ExpectFails()      //<------ new
     .OnMessage(new ExampleMessage());
```

and 

```
Test.Handler<ExampleHandler>()
    .ExpectDoesNotFail()       //<------ new
    .OnMessage(new ExampleMessage());
```

Currently, `Test.Handler` will re-throw any exceptions that occur in the Handler, so I can work around this as described here: http://stackoverflow.com/questions/18634647/nservicebus-3-2-testing-expected-failure, but it would be nice if testing error handling was possible with the same fluent syntax.

**Comment by danielmarbach** [comment]
@ppittle thanks for the suggestion. Out of curiosity: Since you are using bleeding edge, is there a reason why you stay with the older Test syntax instead of using the testable contexts directly? With the testable contexts you could just use normal AAA style tests with something like this

```
            var context = new TestableMessageHandlerContext();
            var handler = new SendingHandler<Send1>();

            Assert.ThrowsAsync<InvalidOperationException>(() => handler.Handle(new TestMessage(), context));
```

**Comment by ppittle** [comment]
@danielmarbach - Because this is the first time I'm seeing `TestableMessageHandlerContext()` :)

tl;dr - AAA sample is not used in any of the sample on the documentation page and is only shown on the samples page.

When I was first reading the documentation on [Testing](http://docs.particular.net/nservicebus/testing/) I glanced at the text briefly, and then went straight to the code samples.  Basically, this is how I read the page:

![image](https://cloud.githubusercontent.com/assets/1190907/18083188/edeaaf02-6e56-11e6-8b13-fb9aa42b1e6a.png)

Because a) the code sample for AAA style was buried in a link [Unit Testing NServiceBus 6 sample](http://docs.particular.net/samples/unit-testing/), yet immediately below was example code.  Frankly, after seeing the sample I couldn't believe how easy it would be to write tests so I immediatly alt-tab'd to Visual Studio to try it.  I've come back to the documentation page a few times to check on how to write tests for Sagas but because everything I needed was on that page, I never really looked at the sample page.

**Comment by danielmarbach** [comment]
Really valuable feedback! I'll raise a docs issue

I glance through the code and such an expectation would be simply to
implement. Which would allow us to also update that SO post with a better
example

**Comment by danielmarbach** [comment]
@ppittle made a quick spike is that what you would like?

https://github.com/Particular/NServiceBus.Testing/pull/91

**Comment by danielmarbach** [comment]
Raised an issue https://github.com/Particular/docs.particular.net/issues/1829

**Comment by ppittle** [comment]
@danielmarbach - Method signatures look great!  AAA probably doesn't need this, as it should already be built into the test frameworks.

Actually, [ExpectNotFail.cs](https://github.com/Particular/NServiceBus.Testing/pull/91/commits/c95f7c8127307aa8078df71cae703040c08606db#diff-325157b0da94f97bf43da2822ddadbcfR18) looks not quite right. Not sure if I'm reading this right, but exception message might not be correct?  I would think it would say something like 
    "Expected no exception to be thrown when processing message, but an exception was thrown"

Also, to use ExpectNotFail() do you need to provide the exception type as a generic type parameter, ie `ExpectNotFail<ArgumentException>()`?   Maybe add an overload that allows you to easily say no exception should be thrown.  This method could possibly be empty; if an exception is thrown the test will short circuit anyway, but it helps to better express intent when writing tests where a Handler should explicitly handle exceptions.

**Comment by danielmarbach** [comment]
@ppittle I didn't finish the ExpectNotFail stuff. I stopped when I got the ExpectFail working

**Comment by andreasohlund** [comment]
@danielmarbach since https://github.com/Particular/docs.particular.net/issues/1829 is done is there more we have to do here given that AAA is the way to go and would not require us to do anything?

**Comment by danielmarbach** [comment]
@andreasohlund are you saying we should not be adding an obviously missing API to the current testing syntax? Like I showed in the PR it would be simple enough and complete the API. But I agree though it is probably not high priority

**Comment by andreasohlund** [comment]
Sorry missed that, you are saying that 

`var ex = Assert.ThrowsAsync<Exception>(async () => myHandler.Handle(...))`

is not sufficient?

**Comment by danielmarbach** [comment]
The argument from @ppittle was that currently there is no ExpectFails method available which would be handy. I agree that AAA syntax allows addressing this like you showed. But we are lacking an official roadmap where we are heading towards with the testing API

**Comment by andreasohlund** [comment]
Since doco shows AAA as the preferred option https://docs.particular.net/samples/unit-testing/ I'd say we are close enough to close this one as won't fix?

**Comment by danielmarbach** [comment]
Wouldn't it then be fair to note in the doco that the FluentTestingAPI will no longer be extended because the new AAA style should be used. Because the doco does not really show that it is the preferred option. I think we would need to make a stronger statement

**Comment by andreasohlund** [comment]
Agreed, how about we:

1. Change https://docs.particular.net/nservicebus/testing/#handlers to show AAA syntax intead
2. Move the "old" syntax to a dedicated page with a WARN at the top mentioning that this is no longer the prefered way to test handlers
3. Adjust https://docs.particular.net/samples/unit-testing/ to link to it with "This is the old and deprecated syntax"

Thoughts?

**Comment by danielmarbach** [comment]
@andreasohlund I did some prep work to get this going. Then I realized one of the things we don't provide with the AAA syntax (which the fluent shows) is dealing with timeouts. It is possible to just call the timeout method from the public API perspective. That is a no brainer. I'm talking about asserting a certain timeout period. For example let's imagine you have the following policy:

```
public class OrderPolicy : Saga<OrderPolicyData>, 
    IAmStartedByMessages<SubmitOrder>,
    IHandleTimeouts<SubmitOrder>
{
    public Task Handle(SubmitOrder message, IMessageHandlerContext context)
    {
        return RequestTimeout(context, TimeSpan.FromMinutes(5), message);
    }

    public Task Timeout(SubmitOrder state, IMessageHandlerContext context)
    {
        var processOrder = new ProcessOrder
        {
            CustomerId = Data.CustomerId,
            OrderId = state.OrderId,
            TotalAmount = state.TotalAmount
        };
        return context.Send(processOrder);
    }

    protected override void ConfigureHowToFindSaga(SagaPropertyMapper<OrderPolicyData> mapper)
    {
    }
}
```
It should be at least possible to do something like

```
[Test]
public async Task ShouldSetBuyersRemorsePeriodToFiveMinutes()
{
    var saga = new OrderPolicy
    {
        Data = new OrderPolicyData()
    };
    var context = new TestableMessageHandlerContext();

    var submitOrder= new SubmitOrder
    {
        CustomerId = Guid.NewGuid(),
        OrderId = Guid.NewGuid(),
        TotalAmount = 1000
    };

    await saga.Handle(discountOrder, context)
        .ConfigureAwait(false);

    var timeoutMessage= context.TimeoutMessages[0];
    var timeoutMessageContent = (SubmitOrder) timeoutMessage.Message;

    Assert.AreEqual(TimeSpan.FromMinutes(5), processMessage.timeoutMessage.Within);
    Assert.AreSame(submitOrder, timeoutMessageContent);
}
```

I checked whether it is possible to implement this in the testing lib only. Unfortunately this is not really possible without doing major hacks. I'm proposing that we implement such a functionality for the next major version of the core in the testing fakes directly.

Should we wait with what you proposed until both syntaxes are on par in feature compatibility?

**Comment by danielmarbach** [comment]
Sent in a PR https://github.com/Particular/NServiceBus/pull/4820

**Comment by danielmarbach** [comment]
PR is merged. The outstanding question is:

> Should we wait with what you proposed until both syntaxes are on par in feature compatibility?

thoughts @Particular/nservicebus-maintainers ?

**Comment by bording** [comment]
@danielmarbach I'd say we go ahead with what @andreasohlund proposed.

**Comment by danielmarbach** [comment]
I will see if I have capacity next week.

**Comment by danielmarbach** [comment]
Sent in https://github.com/Particular/docs.particular.net/pull/2971

**Comment by danielmarbach** [comment]
Pr merged. Closing
